ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -760,6 +760,41 @@ void seal() throws StoreException {
     persistBloomFilter();
   }
 
+  /**
+   * @return index value of last PUT record in this index segment. Return {@code null} if no PUT is found
+   */
+  IndexValue getIndexValueOfLastPut() throws StoreException {
+    IndexValue indexValueOfLastPut = null;
+    if (sealed.get()) {
+      ByteBuffer readBuf = serEntries.duplicate();
+      int numOfIndexEntries = numberOfEntries(readBuf);
+      NavigableSet<IndexValue> values = new TreeSet<>();
+      for (int i = 0; i < numOfIndexEntries; i++) {
+        StoreKey key = getKeyAt(readBuf, i);
+        values.clear();
+        getAllValuesFromMmap(readBuf, key, i, numOfIndexEntries, values);
+        for (IndexValue indexValue : values) {
+          // FLAGS_DEFAULT_VALUE means PUT record
+          if (indexValue.getFlags() == IndexValue.FLAGS_DEFAULT_VALUE && (indexValueOfLastPut == null
+              || indexValue.compareTo(indexValueOfLastPut) > 0)) {
+            indexValueOfLastPut = indexValue;
+          }
+        }
+      }
+    } else {
+      for (Map.Entry<StoreKey, ConcurrentSkipListSet<IndexValue>> entry : index.entrySet()) {
+        for (IndexValue indexValue : entry.getValue()) {
```

Comment:
Basically your concern is when iterating on an entry, the `index` may have changed which makes last PUT invalid. This should not happen because we always check Journal first and Journal covers most recent two index segments. Hence, when we start to check index segment (it is very likely sealed but not guaranteed, seal operation is performed by background persistor), we can guarantee that this index segment being checked is not an active one, that is, no more entries will be added into this index segment. So it is safe to iterate over all entries in this index segment and validity of last PUT still holds